improver_tests/convection/test_ConvectionRatioFromComponents.py

```diff
@@ -0,0 +1,173 @@
+            units="m s-1",
+            spatial_grid="equalarea",
+            grid_spacing=96900.0,
+            domain_corner=(-1036000.0, -1158000.0),
+            attributes=UK_ATTRIBUTES,
+        )
+    )
+    cubes.append(
+        set_up_variable_cube(
+            data.copy(),
+            name="dynamic_precipitation_rate",
+            units="m s-1",
+            spatial_grid="equalarea",
+            grid_spacing=96900.0,
+            domain_corner=(-1036000.0, -1158000.0),
+            attributes=UK_ATTRIBUTES,
+        )
+    )
+    return cubes
+
+
+@pytest.mark.parametrize("grid_fixture", ["global_grid", "uk_grid"])
+def test_basic(request, grid_fixture):
+    """Ensure Plugin returns object of correct type and meta-data and that
+    no precip => masked array"""
+    grid = request.getfixturevalue(grid_fixture)
+    expected_array = np.ma.masked_all_like(grid[0].data)
+    result = ConvectionRatioFromComponents()(grid.copy())
+    assert isinstance(result, iris.cube.Cube)
+    assert_allclose(result.data, expected_array)
+    assert result.attributes == grid[0].attributes
+    assert result.long_name == "convective_ratio"
+    assert result.units == "1"
+
+
+# These tuples represent one data point. The first two values are the convective and
+# dynamic precipitation rate respectively. The last value is the expected result.
+@pytest.mark.parametrize(
+    "data_con_dyn_out",
+    [
+        (1.0, 0.0, 1.0),
+        (1.0, 1, 0.5),
+        (0.0, 1.0, 0.0),
+        (0.9e-9, 0.0, np.inf),
+        (1.1e-9, 0.0, 1.0),
+        (0.9e-9, 0.9e-9, 0.5),
+    ],
+)
+@pytest.mark.parametrize("grid_fixture", ["global_grid", "uk_grid"])
+def test_data(request, grid_fixture, data_con_dyn_out):
+    """Test that the data are calculated as expected for a selection of values
+    including either side of the minimum precipitation rate tolerance 1e-9 m s-1"""
+    grid = request.getfixturevalue(grid_fixture)
+    for i in range(2):
+        grid[i].data[0, 0] = data_con_dyn_out[i]
+    expected_array = np.ma.masked_all_like(grid[0].data)
+    if np.isfinite(data_con_dyn_out[2]):
+        expected_array[0, 0] = data_con_dyn_out[2]
+    result = ConvectionRatioFromComponents()(grid)
+    assert_allclose(result.data, expected_array)
+    # assert_allclose doesn't check masks appropriately, so check separately
+    assert_equal(result.data.mask, expected_array.mask)
+
+
+def test_bad_name(request):
+    """Test we get a useful error if one of the input cubes is incorrectly named."""
+    grid = request.getfixturevalue("uk_grid")
+    grid[0].rename("kittens")
+    with assert_raises_regex(
+        ValueError, "Cannot find a cube named 'convective_precipitation_rate' in "
+    ):
+        ConvectionRatioFromComponents()(grid)
+
+
+def test_bad_units(request):
+    """Test we get a useful error if the input cubes have non-SI units."""
+    grid = request.getfixturevalue("uk_grid")
+    [c.convert_units("mm h-1") for c in grid]
```

Comment:
Codacy found an issue: [Expression "[c.convert_units('mm h-1') for c in grid]" is assigned to nothing](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5973665)